### DISCUSSION TITLE
Debugging ksh, trying to find printf %T bug and also supported format

### DISCUSSION THREAD

**Discussion by kshji** [ksh93/ksh#487] [Q&A]
Source code include **message** function calls, how I can setup show that I get message() output? And also debug() output?
I have compiled ksh using to get debug version of ksh
`bin/package make CCFLAGS="-O -g" debug=1`

I'm just looking libast/tm/tmxdate.c to find bug. the message output helps to read parser.
I have not found full manual with some examples when using excellent printf '%T'.

There is bug 
* printf '%(%a)T\n' 'final day june 2001' works fine, result is correct
* printf '%(%Y-%m-%d)T\n' 'last month'
* printf "%(%Y-%m-%d)T\n" "final last 10 month" - result correct
* printf '%T\n' '10 month ago' works fine, result is correct
* printf '%T\n' '20 month ago' **give wrong answer**
* printf "%(%Y-%m-%d)T\n" "final last 20 month"  - **give wrong answer**
** if you split: take 10 and 10 and 5 months, you got correct answer
* printf '%(%Y-%m-%d)T\n' '3rd wednesday july 2022' - give correct answer
* printf '%(%Y-%m-%d %H:%M)T\n' 'last workday' - give correct answer  (Mon-Fri are workdays)
I have tested cron format, nice:
* printf '%(%Y-%m-%d %H:%M)T\n' '5 6 1,15 * 1' - result is 2022-08-01 06:05

All reserved words (supported) in file libast/tm/tmdata.c

I found from source that you can  use also  sort format like 2M , 10H, ... 2Y, 2h and so on works also. Result is not same as "2 month".
* today is 2022-07-06
* printf '%(%Y-%m-%d)T\n' '10 month ago' - give correct result  2021-09-01 
* printf '%(%Y-%m-%d)T\n' 2M   - give correct result 2022-09-06 - exactly 2 month
* printf '%(%Y-%m-%d)T\n' '2 month'  -  give correct result 2022-09-01 = first day of 2 month forward
* printf "%(%Y-%m-%d %H:%M)T\n" "M-2" -  give result 2022-02-06 17:29 = 2nd month, not -2 months
* But how I can use '2M' with some ??? but ago = result have to be currentdate - 2 months = 2022-05-06 
** I have tried -, ago,past ... but always I got 2022-09-06 , not 2022-05-06

**Comment by McDutchie** [comment]
First of all, tervetuola/welcome! I'm glad someone is trying to work on this as no one has had time or motivation so far.  See also #182 for more breakage; that issue would be a good place to discuss and test possible fixes.

You asked how to get those `message()` calls to show. That debug macro is defined in ast.h, here: https://github.com/ksh93/ksh/blob/9c037b0458dc44b37f7f8ca773b307e76b41e6b1/src/lib/libast/include/debug.h#L38-L41 So it is conditional upon either `DEBUG` or `_BLD_DEBUG` being defined. So you need to add either `-DDEBUG` or `-D_BLD_DEBUG` to `CCFLAGS`.

Then, whether it shows a message is also dependent on `error_info.trace`, as you can see above. The [error.3 manual page](https://github.com/ksh93/ksh/blob/9c037b0458dc44b37f7f8ca773b307e76b41e6b1/src/lib/libast/man/error.3#L240-L283) documents how to make that happen:
```
EXTENDED DESCRIPTION
       <debug.h> provides debugging message macros when DEBUG or  _TRACE_  are
       defined  (_TRACE_  is  defined  by makerules when CCFLAGS contains -g).
       All of the macros expand to nothing when both DEBUG and _TRACE_ are not
       defined.   Otherwise  debug  expands its arg and libmessage and message
       call liberror and error  respectively  if  error_info.trace<0.   Notice
       that  libmessage and message are macro hacks that require double paren-
       theses ((...)) around the arguments.

EXAMPLE
       To enable debugging message level  -3,  library  messages,  and  system
       errno text for all commands:

              export ERROR_OPTIONS="trace=3 library system"
```
Actually, I think there's a mistake in that: `_TRACE_` does not exist anywhere. I think they mean `_BLD_DEBUG`. Sure enough, a quick search in [ksh93-history](https://github.com/ksh93/ksh93-history) shows that `_TRACE_` was changed to `_BLD_DEBUG` in debug.h on or before 2000-10-31 ksh93k; they forgot to update the manual. That fix will go into the next round of minor cleanups. The build system also no longer automatically adds either of these flags (probably due to nmake being removed); I should either find a way to fix that or update the documentation.

Anyhow, in summary: compiling with either `-DDEBUG` or `-D_BLD_DEBUG` and then doing `export ERROR_OPTIONS="trace=3 library system"` will get those `message()` calls to show.

**Comment by McDutchie** [comment]
Well, would you look at that... passing `debug=1` to `bin/package` *will* actually get the system to define `_BLD_DEBUG` but *only* if you don't have your own `CCFLAGS` defined! If you do, they will override that flag and you have to add it manually. That's because the Mamfiles only define the debug flags as the default value for `CCFLAGS` which is simply overwritten if you pass your own: https://github.com/ksh93/ksh/blob/9c037b0458dc44b37f7f8ca773b307e76b41e6b1/src/lib/libast/Mamfile#L14 (note that the `-debug-symbols` MAM variable is set by mamake if the `-G`/`--debug-symbols` option was passed to it).

Anyway, the following will work to debug tmxdate.c:
```
$ unset CCFLAGS
$ bin/package make debug=1
$ export ERROR_OPTIONS="trace=1"
$ arch/*/bin/ksh -c 'printf %T\\n "4 years ago"'
printf: debug: AHA tmxdate 2009-03-06
printf: debug: AHA#233 state=| set=| '4 years ago'
printf: debug: AHA#712 n=4 w=1 u=' ' f=-1 t=" years ago"
printf: debug: AHA#841 n=4 u="years ago"
printf: debug: AHA#1090 s="years ago"
printf: debug: AHA#1108 n=4 q='y'
printf: debug: AHA#1335 n=4 j=61 f=-1 state=|LAST|NEXT|
printf: debug: AHA#1357 disambiguate LAST s=''
printf: debug: AHA#1372 disambiguate LAST k=-1
printf: debug: AHA#1378 f=-1 n=-4 i=1 j=61 k=-1 l=0 m=-4 state=|LAST|NEXT|
printf: debug: AHA#1693 DAY
printf: debug: AHA#233 state=| set=|LAST| ''
Mon Jan  1 00:00:00 CET 2018
```

**Reply by kshji** [reply]
Thank you. I hope that this answer helps all others who are interested to look ksh93, debug ksh93. 
I'm just retired, so some raining day I start to look more why the ex. 25 month ago give wrong answer.

**Comment by phidebian** [comment]
I was asking for debug technic in my first post, and there are good start here :)

**Reply by phidebian** [reply]
@kshji  I guess we can ignore things like M-2 etc as I can't find any docco about it, and gnudate is bogus on that anyway, so I guess this M+xyz M-xyz are kind of HP-41C secret opcode?

I got this with date(1)

    TC$ date --date="now" +"%Y-%m-%d %H:%M:%S"                
    2022-10-14 11:53:03
    
    TC$ date --date="M-2" +"%Y-%m-%d %H:%M:%S"      
    2022-10-13 16:00:00
    
    TC$ date --date="M-1" +"%Y-%m-%d %H:%M:%S"
    2022-10-13 15:00:00
    
    TC$ date --date="M-2" +"%Y-%m-%d %H:%M:%S"
    2022-10-13 16:00:00
    
    TC$ date --date="M-3" +"%Y-%m-%d %H:%M:%S"
    2022-10-13 17:00:00

M-xyz sometime touch the Month, sometime the Hours, look buggy to me.
